stateproof/builder.go

```diff
@@ -201,7 +364,19 @@ func (spw *Worker) handleSig(sfa sigFromAddr, sender network.Peer) (network.Forw
 				sfa.Round, proto.StateProofInterval)
 		}
 
-		builderForRound, err = spw.makeBuilderForRound(sfa.Round)
+		if sfa.Round > latest {
+			// avoiding an inspection in DB in case we haven't reached the round.
+			// Avoiding disconnecting the peer, since it might've been sent to this node while it recovers.
+			return network.Ignore, fmt.Errorf("handleSig: latest round is smaller than given round %d", sfa.Round)
+		}
+
+		// We want to save the signature in the DB if we know we generated it. However, if the signature's source is
+		// external, we only want to process it if we know for sure it meets our broadcast policy.
+		if sender != nil && !spw.meetsBroadcastPolicy(sfa, latestHdr.Round, &proto, stateProofNextRound) {
+			return network.Ignore, nil
+		}
+
+		builderForRound, err = spw.loadOrCreateBuilderWithSignatures(sfa.Round)
```

Comment:
The builder's cache is trimmed periodically via `Worker.trimBuildersCache`, so there are at most 5 builders in memory at any given point (barring times between trims when an extra 6th builder might be present).
4 of the builders will be the earliest builders [`NextStateProofRound, ..., NextStateProofRound+Interval*3`] and the 5th one is the latest round builder, which is replaced every new interval.

As for the transactions being broadcasted, you can notice we break after the first transaction gets rejected by the local pool, thus no more transactions will be built and broadcasted at that same round:
```
	if err != nil {
		spw.log.Warnf("spw.tryBroadcast: broadcasting state proof txn for %d: %v", rnd, err)
		// if this StateProofTxn was rejected, the next one would be rejected as well since state proof should be added in
		// a sequential order
		break
	}
```
Only the first transaction is accepted, we try to build and add the next StateProof transaction (allowing for a single block with up to 4 StateProof transactions).

Reply:
Currently, `tryBroadcast` breaks only if the state proof is rejected by the local pool (the advantage is that the relay can create the proof and have it ready. This allows the replay to send several statproofs in one block). For simplicity, we can break the loop on every condition

Reply:
Thank you @algoidan and @Aharonee for the clarification. I agree with everything you described.

So, is there a legitimate reason why a node should be getting signatures for stateproofs which are going to be deleted in `Worker.trimBuildersCache`? (except for the onese < nextStateproofRound)

In other words, since no node will create builders, sign and broadcast signatures for these rounds (> threshold and < latest), then no such signature will be handled to trigger the creation of a builder which then will be trimmed. 

If so, will it be helpful to log a warning when builders are getting trimmed? Specifically, only report warning for trimmed builders > threshold, but not for the trimmed builders which already had the stateproofs accepted to the block.

Reply:
Every Relay will have up to `buildersCacheLength` (might be +1 for 1-2 rounds until `trimBuildersCache` is triggered ) builders stored in memory. In correlation to this, every node will send signatures that correspond to those builders. Also, notice that  `handleSig` function, reject signatures if they don't fit this policy 
`if sender != nil && !spw.meetsBroadcastPolicy(sfa, latestHdr.Round, &proto, stateProofNextRound) `
So one can't spam relay with nonrelevant signatures.

What benefit do you think we will get from writing a log when a builder gets trimmed? (Since trimming builder is the expected behavior)

Reply:
Trimming builders for rounds > threshold and < latest is not expected behavior. Excessive cases of this is an indication of something wrong. 
As we discussed, only expected cases of these will be a node catching up that might trigger this, which is expected to be very rare, so will have no impact on the logs. 
The impact on the logs is considerable when something is seriously wrong and we need to be alerted about.